algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+
+structure isomorphism (α : Type u) (β : Type v) [comm_ring α] [comm_ring β] :=
+(f : α → β) (hf : is_hom f)
+(g : β → α) (hg : is_hom g)
+(hfg : ∀ x, f (g x) = x)
+(hgf : ∀ x, g (f x) = x)
+
+infix `≅`:70 := isomorphism
+
+@[simp] lemma quotient.lift_beta {α : Sort u} {β : Sort v} [s : setoid α] (f : α → β) (h : ∀ (a b : α), a ≈ b → f a = f b) (x : α):
+quotient.lift f h (quotient.mk x) = f x := rfl
+
+@[simp] lemma quotient.lift_on_beta {α : Sort u} {β : Sort v} [s : setoid α] (f : α → β) (h : ∀ (a b : α), a ≈ b → f a = f b) (x : α):
+quotient.lift_on (quotient.mk x) f h = f x := rfl
+
+noncomputable def first_isom (α : Type u) (β : Type v) [comm_ring α] [comm_ring β] (f : α → β) [is_hom f] :
+(α / (is_hom.ker f)) ≅ (is_hom.im f) :=
+{ f := λ x, quotient.lift_on x (λ x, ⟨f x, x, rfl⟩ : α → is_hom.im f) (λ x y hxy, subtype.eq $ calc
+    f x = f (y + (x - y)) : by norm_num
+    ... = f y + f (x - y) : is_hom.map_add f
+    ... = f y             : by simp [has_equiv.equiv, setoid.r] at hxy; simp [hxy]),
+  hf :=
+    { map_add := λ x y, quotient.induction_on₂ x y (λ m n, by simp [is_hom.map_add f]; refl),
+      map_mul := λ x y, quotient.induction_on₂ x y (λ m n, by simp [is_hom.map_mul f]; refl),
+      map_one := by simp [is_hom.map_one f]; refl },
+  g := λ ⟨y, hy⟩, ⟦classical.some hy⟧,
+  hg :=
+    { map_add :=
+        begin
+          intros x y,
+          cases x with x hx,
+          cases y with y hy,
+          simp [first_isom._match_1],
+          have hm := classical.some_spec hx,
+          have hn := classical.some_spec hy,
+          have hmn := classical.some_spec (subring.add_mem hx hy),
+          simp [is_hom.map_add f, is_hom.map_neg f, hm, hn, hmn]
+        end,
+      map_mul :=
+        begin
+          intros x y,
+          cases x with x hx,
+          cases y with y hy,
+          simp [first_isom._match_1],
+          have hm := classical.some_spec hx,
+          have hn := classical.some_spec hy,
+          have hmn := classical.some_spec (subring.mul_mem hx hy),
+          simp [is_hom.map_add f, is_hom.map_neg f, is_hom.map_mul f, hm, hn, hmn]
+        end,
+      map_one :=
+        begin
+          apply quotient.sound,
+          have h := classical.some_spec (subring.one_mem $ is_hom.im f),
+          simp [is_hom.map_add f, is_hom.map_neg f, h, is_hom.map_one f]
+        end },
+  hfg := λ ⟨x, hx⟩, subtype.eq (by simp [first_isom._match_1]; simpa using classical.some_spec hx),
+  hgf :=
+    begin
+      intro x,
+      cases is_ideal.coset_rep x with y hy,
+      rw ← hy,
+      simp [first_isom._match_1],
+      have hz := @classical.some_spec _ (λ z, f z = f y) ⟨y, rfl⟩,
+      simp [is_hom.map_add f, hz, is_hom.map_neg f]
+    end
+}
+
+local infix `^` := monoid.pow
+
+def nilpotents (α : Type u) [comm_ring α] : set α := { x | ∃ n, x^(nat.succ n) = 0 }
+def is_unit {α : Type u} [comm_ring α] (x : α) := ∃ y, x * y = 1
+def nonunits (α : Type u) [comm_ring α] : set α := { x | ¬∃ y, x * y = 1 }
+
```

Comment:
Is there an argument for also defining is_nilpotent, units, and is_nonunit? One could imagine using them -- especially units, which is canonically an abelian group under multiplication.